### PR TITLE
#58 set health status on startup & shutdown

### LINKED ISSUE #58: Set health status on startup and shutdown
If a shutdown is detected, the health status of the component-operator component should be set to `unavailable` and the health of all other components to `unknown`.
On startup, the health of all components should be detected and set.

### PR DESCRIPTION
**Pull request by jelemux**
Resolves #58

### CODE REVIEW

File: pkg/health/shutdownHandler.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+const componentOperatorName = "k8s-component-operator"
+
+type ShutdownHandler struct {
+	repo componentRepo
+}
+
+func NewShutdownHandler(client componentClient) *ShutdownHandler {
+	return &ShutdownHandler{repo: &defaultComponentRepo{client: client}}
```
Comment (nhinze23) [inline pkg/health/shutdownHandler.go:20]: The componentRepo could be created in the main method and used here and in the Manager.

File: pkg/health/shutdownHandler.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (s *ShutdownHandler) Start(ctx context.Context) error {
+	logger := log.FromContext(ctx).WithName("shutdown handler")
+	logger.Info("waiting for shutdown")
```
Comment (nhinze23) [inline pkg/health/shutdownHandler.go:25]: I think this log could be confusing. With structured logging you will see that it comes from this handler but at the operator start this is not really needed.

File: pkg/health/shutdownHandler.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (s *ShutdownHandler) handle(ctx context.Context) error {
+	// set component operator health status to unavailable
+	componentOperator, err := s.repo.get(ctx, componentOperatorName)
+	if err != nil {
+		return fmt.Errorf("failed to get component for %q: %w", componentOperatorName, err)
+	}
+
+	err = s.repo.updateCondition(ctx, componentOperator, v1.UnavailableHealthStatus, noVersionChange)
```
Comment (nhinze23) [inline pkg/health/shutdownHandler.go:41]: I think we have to refactor this method in future, so we don't have to set the status and version at the same time.

File: pkg/health/shutdownHandler.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	logger.Info("shutdown detected, handling health status")
+
+	// context is done, we need a new one
+	ctx = context.WithoutCancel(ctx)
+	return s.handle(ctx)
+}
+
+func (s *ShutdownHandler) handle(ctx context.Context) error {
+	// set component operator health status to unavailable
+	componentOperator, err := s.repo.get(ctx, componentOperatorName)
```
Comment (nhinze23) [inline pkg/health/shutdownHandler.go:36]: In dev mode with `helm-apply` this will through an error.

File: pkg/health/shutdownHandler_test.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			wantErr: assert.NoError,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &ShutdownHandler{
+				repo: tt.repoFn(t),
+			}
+			ctx, cancel := context.WithCancel(testCtx)
+			cancel()
```
Comment (nhinze23) [inline pkg/health/shutdownHandler_test.go:185]: Oh nice. I have to use this in the dogu-operator. Just cancel the context before calling `Start` instead using a timer. :laughing:

File: pkg/health/shutdownHandler.go (hunk 3 above)
Comment (jelemux) [reply pkg/health/shutdownHandler.go:41]: Yeah, I found it ugly as well. Had to resolve quite a bit of conflicts when I merged it from develop.

File: pkg/health/shutdownHandler.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	<-ctx.Done()
+	logger.Info("shutdown detected, handling health status")
+
+	// context is done, we need a new one
+	ctx = context.WithoutCancel(ctx)
+	return s.handle(ctx)
+}
+
+func (s *ShutdownHandler) handle(ctx context.Context) error {
+	// set component operator health status to unavailable
```
[the remark this answers is not in the document]
Comment (jelemux) [reply pkg/health/shutdownHandler.go:35]: True, I'll get it in the list anyway. I guess it's due to the way I worked through the tasks. I'll change it.

File: pkg/health/shutdownHandler.go (hunk 2 above)
Comment (jelemux) [reply pkg/health/shutdownHandler.go:25]: Ok, I can remove it

File: pkg/health/shutdownHandler.go (hunk 1 above)
Comment (jelemux) [reply pkg/health/shutdownHandler.go:20]: Hm, I didn't like having to hand too many objects through all those methods. That's why I settled for the client as that is used by other parts of the application as well.

File: pkg/health/shutdownHandler.go (hunk 2 above)
Comment (jelemux) [reply pkg/health/shutdownHandler.go]: Didn't remove it, but made it less confusing

File: pkg/health/shutdownHandler.go (hunk 4 above)
Comment (jelemux) [reply pkg/health/shutdownHandler.go]: This is a general problem, we'll create a bug report.

**Comment by nhinze23** [comment]
Nice work :+1:  lgtm

**Pull request merged**